### DISCUSSION TITLE
Create geometries from primitives and boolean operators

### DISCUSSION THREAD

**Discussion by oberstet** [thliebig/openEMS-Project#138] [Q&A Usage] [answered]
Hello all,

I've got a specific question rgd geometry creation: is it possible to define geometries from boolean operations on object primitives?

E.g. can I intersect a CylindricalShell with a Box?

Maybe easier to describe by what I want to achieve .. maybe there is a simple way .. otherwise it'll get tedious.

So I want to create geometries in openEMS using Python modeling the following PCB antenna which is just a non-optimized "sketch" in KiCad I created:

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/5b9510e1-521f-47da-8a2f-3fe552575ab1)

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/4028ee80-c6d0-4640-b246-2e5f4e1ac34e)

I want to have a parametrized geometry model in Python, currently, I've got this in openEMS:

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/f44fdb19-8410-41e4-93a4-33e947ef5601)

-----

Currently, it uses Wires between the helix turning points, but I want to use Boxes (I can do that) .. BUT .. please have a look at this problem:

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/203c3829-c089-4491-b90f-7f21c364f3d3)

So the Wires extend _into_ the inner space of the CylindricalShells ... which is obviously wrong ... the Wire bodies must "merge" with the CylindricalShells.

I use the CylindricalShells to model Vias.

**Comment by LubomirJagos** [comment, marked as answer]
Hi, I haven't seen any boolean operations in openEMS interface, but what I have experience you have these options:
    - create parametrized model using python which generates STL using some 3rd party library and then import generated model using PolyhedronReader
    - using primitives using python functions with set priorities, at the place of object intersections the object with higher priority is taken, so if same materials intersect you don't have to do anything, for substraction you have to define material with er=1, mu=1 and give it highest priority therefore it will have same effect like cutting out these parts of model
    - you can use fcad_pcb (https://github.com/realthunder/fcad_pcb) in FreeCAD and import your KiCAD sketch into FreeCAD and export it as STL, FreeCAD using python for scripting so this way you can directly use your KiCAD model
    
Hope this helps.

**Reply by oberstet** [reply]
Thank you so much for replying, helping!

> I haven't seen any boolean operations in openEMS interface

Ok, I see:(

> but what I have experience you have these options:

Thanks a lot for describing your workflow! This help judging my options, and fcad_pcb for transporting to KiCad: that would be ideal. Auto-optimize antenna parameters in Python, then export KiCad footprint for antenna to reuse.

I am still piecing together one for what I'd like to do. I have a ton of Python/SW know-how/experience. But pretty much zero in this advanced RF stuff. I want to get something running .. even if only "close" to what I need. Currently, I'm struggling with meshes .. will post another Q;)

**Comment by samerps** [comment]
you can use Blender to create complex geometries, and then export them as STLs

**Reply by oberstet** [reply]
thanks for chiming in! all of these tools are new to me ... kicad, freecad, openems and blender also .. just had a quick look rgd "APIs", turns out blender has _also_ a python API =) so it seems what you suggest is yet another option! btw: I love your antenna movies on YT!! as a newbie with only basic knowledge of RF, Maxwell and all that, e.g. developing an intuition for how an inverted F works, and specifically how a "trace shortened to GND" can ever act like a capacitor, was gold!! love it. well, I'm not yet even at the stage where I could verify this planar helical thingy can be made working at all at the intended frequencies;)
